### PR TITLE
deactivate endpoints unless admin token set

### PR DESCRIPTION
**Pull request by ameijer**
## Description

Fixes CVE-2026-44300 by failing closed on admin-protected HTTP endpoints when `ADMIN_TOKEN` is unset.

- `adminAuthMiddleware` now returns HTTP 503 (with `Cache-Control: no-store`) when `ADMIN_TOKEN` is not configured, instead of warn-and-allow.
- `GET /helmValues` is now protected by `adminAuthMiddleware` (it returns decoded `HELM_VALUES`, which may contain secrets).

## Related Issues

Fixes CVE-2026-44300

## User Impact

**Breaking change.** Deployments without `ADMIN_TOKEN` lose access to:

- `POST /serviceKey`
- `GET /helmValues`
- `GET /cloudCost/rebuild`, `GET /cloudCost/repair`
- `GET /cloud/config/export`, `GET /cloud/config/enable`, `GET /cloud/config/disable`, `GET /cloud/config/delete`

Set `ADMIN_TOKEN` before using these endpoints. The [opencost-helm-chart](https://github.com/opencost/opencost-helm-chart) supports this via `opencost.exporter.adminToken` (currently opt-in / disabled by default).

If opencost-ui calls `/helmValues`, default installs without `ADMIN_TOKEN` will see an error until the token is configured.

## Testing

- `go test ./pkg/costmodel/ -run TestAdminAuth`
- Verified middleware returns 503 + `Cache-Control: no-store` when `ADMIN_TOKEN` is unset
- Verified 401/403 behavior unchanged when `ADMIN_TOKEN` is set

### CODE REVIEW

File: pkg/costmodel/router.go [hunk 1]
```diff
@@ -131,15 +131,14 @@ func ParsePercentString(percentStr string) (float64, error) {
...
+// adminAuthMiddleware wraps a handler and requires a Bearer token matching ADMIN_TOKEN.
+// When ADMIN_TOKEN is not set, returns 501 — the endpoint is disabled until configured.
 // When ADMIN_TOKEN is set, returns 401 if the Bearer token is missing or 403 if it does not match.
 func adminAuthMiddleware(next httprouter.Handle) httprouter.Handle {
 	return func(w http.ResponseWriter, r *http.Request, ps httprouter.Params) {
 		adminToken := env.GetAdminToken()
 		if adminToken == "" {
-			log.DedupedWarningf(5, "Admin token (ADMIN_TOKEN) not configured; write operations are unauthenticated")
-			next(w, r, ps)
+			http.Error(w, "Admin token is required to activate this endpoint; set the ADMIN_TOKEN environment variable", http.StatusNotImplemented)
```
Comment (peatey) [inline pkg/costmodel/router.go]: **HTTP 501 is a questionable status choice here.** Per RFC 9110, 501 means "the server does not support the functionality required to fulfill the request", and notably **501 responses are heuristically cacheable by default** (§15.6.2). A cache or proxy in front of OpenCost could keep serving the "disabled" response even after an operator sets `ADMIN_TOKEN`.

`403 Forbidden` or `503 Service Unavailable` would be more conventional for "configured off". If 501 is kept deliberately (to distinguish "endpoint disabled" from "auth failed"), consider adding `Cache-Control: no-store` to the response:

```go
w.Header().Set("Cache-Control", "no-store")
http.Error(w, "Admin token is required to activate this endpoint; set the ADMIN_TOKEN environment variable", http.StatusNotImplemented)
```

---
_Generated by [Claude Code](https://claude.ai/code)_

File: pkg/costmodel/router.go [hunk 2]
```diff
@@ -591,7 +590,7 @@ func Initialize(router *httprouter.Router, additionalConfigWatchers ...*watcher.
 	router.GET("/installNamespace", a.GetInstallNamespace)
 	router.GET("/installInfo", a.GetInstallInfo)
 	router.POST("/serviceKey", adminAuthMiddleware(a.AddServiceKey))
-	router.GET("/helmValues", a.GetHelmValues)
+	router.GET("/helmValues", adminAuthMiddleware(a.GetHelmValues))
```
Comment (peatey) [inline pkg/costmodel/router.go:595]: Gating `/helmValues` is the right call security-wise — `GetHelmValues` decodes and returns the entire `HELM_VALUES` env var (which can contain cloud credentials and other secrets set via Helm values) with `Access-Control-Allow-Origin: *`, previously with no auth at all.

Two follow-ups worth checking, though:

1. Does opencost-ui call `/helmValues`? If so, it will now render an error for default installs (no `ADMIN_TOKEN` set).
2. This plus the fail-closed middleware change means every existing deployment without `ADMIN_TOKEN` loses `/serviceKey`, `/cloudCost/rebuild`, `/cloudCost/repair`, all `/cloud/config/*` endpoints, and `/helmValues`. That's a breaking change that should be spelled out in the PR's "User Impact" section / release notes, with a coordinated [opencost-helm-chart](https://github.com/opencost/opencost-helm-chart) change to expose the token.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: AGENTS.md [hunk 3]
```diff
@@ -147,6 +147,12 @@ just validate-protobuf
 | `MCP_SERVER_ENABLED` | `false` | Enable MCP server |
 | `MCP_HTTP_PORT` | `8081` | MCP server HTTP port |
 
+### Admin Auth
+
+| Variable | Default | Description |
+|----------|---------|-------------|
+| `ADMIN_TOKEN` | (unset) | Bearer token for admin endpoints (`POST /serviceKey`, cloud config mutate/export, `/helmValues`). If unset, those endpoints return HTTP 501. |
```
Comment (peatey) [inline AGENTS.md]: Minor: this description omits `/cloudCost/rebuild` and `/cloudCost/repair`, which are also gated by `adminAuthMiddleware` (see `pkg/costmodel/router.go`). Suggest listing them (or saying "cloud cost rebuild/repair") so the docs match the full set of gated endpoints.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: README.md [hunk 4]
```diff
@@ -1,5 +1,6 @@
 [![License](https://img.shields.io/badge/License-Apache%202.0-blue.svg)](https://opensource.org/licenses/Apache-2.0)
 [![OpenSSF Best Practices](https://www.bestpractices.dev/projects/6219/badge)](https://www.bestpractices.dev/projects/6219)
+[![Known Vulnerabilities](https://snyk.io/test/github/opencost/opencost/badge.svg)](https://snyk.io/test/github/opencost/opencost)
```
Comment (peatey) [inline README.md]: The Snyk badge and the Snyk Secure Developer Program paragraph below are unrelated to the CVE fix. Suggest splitting them into a separate PR — bundling them into a security fix makes backporting/cherry-picking to release branches noisier.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by peatey — COMMENTED** [review]
This PR flips `adminAuthMiddleware` from fail-open to fail-closed (HTTP 501 when `ADMIN_TOKEN` is unset, instead of warn-and-allow) and newly gates `GET /helmValues` behind it.

**The core change is correct, well-tested, and the right security posture.** `GetHelmValues` returns the decoded `HELM_VALUES` env var — which can contain secrets — previously unauthenticated and with `Access-Control-Allow-Origin: *`, so fail-closed is the right default. The constant-time compare and 401/403 split for the configured path are unchanged, and the test updates are coherent (including the "bearer supplied but token unset — still 501" edge case).

Before merging, I'd want:

1. **Breaking-change documentation** — every deployment without `ADMIN_TOKEN` (likely nearly all, since the variable is recent) loses `/serviceKey`, `/cloudCost/rebuild`, `/cloudCost/repair`, `/cloud/config/*`, and `/helmValues`. The PR's "User Impact" and "Testing" sections are empty; this needs a release-notes callout and a coordinated helm-chart change.
2. **A decision on 501 vs 403/503** (or at least `Cache-Control: no-store`) — see inline comment.
3. **The Snyk README additions split into their own PR** — see inline comment.

Minor, non-blocking: the test's env save/restore could use `t.Setenv`, which restores automatically and fails cleanly under `t.Parallel()`.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by cpetersen5 — APPROVED** [review]
